File: C++/Two Pointers/259. 3Sum Smaller.cpp
#include <vector>

class Solution {
public:
    int threeSumSmaller(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int res = 0;
        for (int i = 0; i < nums.size(); i++) {
            twoSum(nums, i, target, res);
        }
        return res;
    }
    
    void twoSum(vector<int>& nums, int index, int target, int& res) {
        int left = index+1, right = nums.size()-1;
        while (left < right) {
            int s = nums[index] + nums[left] + nums[right];
            if (s < target) {
                res += right - left;
                left++;
            }
            else {
                right--;
            }
        }
    }
};

File: Python/Archive/259. 3Sum Smaller.py
from typing import List

class Solution:
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = [0]
        for i in range(len(nums)):
            self.twoSum(nums, i, target, res)
        return res[0]
            
    def twoSum(self, nums, index, target, res):
        left, right = index+1, len(nums)-1
        while left < right:
            s = nums[index] + nums[left] + nums[right]
            if s < target:
                res[0] += right - left
                left += 1
            else:
                right -= 1